### PR TITLE
Add destructuring support to no-shadowed-variable rule

### PR DESCRIPTION
**Pull request by adidahiya**

### CODE REVIEW

File: src/rules/noShadowedVariableRule.ts [hunk 1]
```diff
@@ -31,15 +31,37 @@ class NoShadowedVariableWalker extends Lint.BlockScopeAwareRuleWalker<ScopeInfo,
         return new ScopeInfo();
     }
 
+    public visitBindingElement(node: ts.BindingElement) {
+        // TODO: handle node.dotDotDotToken?
+        var isSingleVariable = node.name.kind === ts.SyntaxKind.Identifier;
```
Comment (leeavital) [inline src/rules/noShadowedVariableRule.ts]: I think this is worth extracting into a utility method. I've seen it at least four times in your recent PRs.

Comment (adidahiya) [inline src/rules/noShadowedVariableRule.ts]: which part? just this one-line check for syntax kind? I disagree; having too many utilities is bad and it's pretty readable here as-is

**Comment by adidahiya** [comment]
waiting on #440 so I can rebase and get `Lint.isBlockScopedBindingElement`

File: src/rules/noShadowedVariableRule.ts [hunk 2]
```diff
@@ -50,38 +72,35 @@ class NoShadowedVariableWalker extends Lint.BlockScopeAwareRuleWalker<ScopeInfo,
...
-    public visitVariableDeclaration(node: ts.VariableDeclaration) {
-        const propertyName = <ts.Identifier> node.name;
-        const variableName = propertyName.text;
+    private handleSingleVariableIdentifier(variableIdentifier: ts.Identifier, isBlockScoped: boolean) {
+        const variableName = variableIdentifier.text;
         const currentScope = this.getCurrentScope();
         const currentBlockScope = this.getCurrentBlockScope();
 
-        // this var is shadowing if there's already a var of the same name in any available scope AND
+        // this const is shadowing if there's already a const of the same name in any available scope AND
```
Comment (adidahiya) [inline src/rules/noShadowedVariableRule.ts]: ugh these should be const

**Comment by adidahiya** [comment]
ready for review

**Comment by adidahiya** [comment]
@leeavital ping for review

File: test/files/rules/no-shadowed-variable.test.ts [hunk 3]
```diff
@@ -75,3 +75,23 @@ function testDoStatement() {
         let foo = 2; // failure
     } while (true);
 }
+
+function testDestructuring() {
+    var x, y, z;
```
Comment (leeavital) [inline test/files/rules/no-shadowed-variable.test.ts]: Can you add some destructuring bindings to make sure they get added to the scope? e.g.

````typescript
var x;
var {y, z} = {y: 2, z: 3}
````

**Comment by leeavital** [comment]
Other than the test, this looks good.

File: test/files/rules/no-shadowed-variable.test.ts (hunk 3 above)
Comment (adidahiya) [inline test/files/rules/no-shadowed-variable.test.ts]: good call

**Pull request merged**